src/java/org/apache/cassandra/config/ValidatedPropertyLoader.java

```diff
@@ -0,0 +1,183 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.config;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.stream.Collectors;
+
+import org.apache.commons.lang3.ArrayUtils;
+
+import org.apache.cassandra.exceptions.ConfigurationException;
+import org.apache.cassandra.utils.FBUtilities;
+import org.yaml.snakeyaml.introspector.FieldProperty;
+import org.yaml.snakeyaml.introspector.Property;
+
+public final class ValidatedPropertyLoader implements Loader
+{
+    private final Loader loader;
+
+    public ValidatedPropertyLoader()
+    {
+        this.loader = new DefaultLoader(ValidatedPropertyLoader::listenablePropertyFactory);
+    }
+
+    @Override
+    public Map<String, Property> getProperties(Class<?> root)
+    {
+        return loader.getProperties(root);
+    }
+
+    private static Property listenablePropertyFactory(Field field)
+    {
+        ListenableProperty<?, ?> listenable = new ListenableProperty<>(new FieldProperty(field));
+        ValidatedByList validatedByList = field.getAnnotation(ValidatedByList.class);
+        ValidatedBy[] validatedByArray;
+        if (validatedByList == null)
+        {
+            if (field.getAnnotation(ValidatedBy.class) == null)
+                return listenable;
+            else
+                validatedByArray = new ValidatedBy[]{ field.getAnnotation(ValidatedBy.class) };
+        }
+        else
+            validatedByArray = validatedByList.value();
+
+        for (ValidatedBy validatedBy : validatedByArray)
+            listenable.addListener(createValidationListener(field, validatedBy));
+
+        return listenable;
+    }
+
+    private static <S, T> ListenableProperty.Listener<S, T> createValidationListener(Field field, ValidatedBy annotation)
+    {
+        Class<?> clazz = FBUtilities.classForName(annotation.useClass(), "validate method");
+        List<Method> matches = Arrays.stream(clazz.getDeclaredMethods())
```

Comment:
Dont we have something in FBUtilties for logic? We could probably move it there for greater good if not there yet.

Reply:
maybe org.apache.cassandra.utils.ReflectionUtils ?

Reply:
Can you share your idea of moving the code to `ReflectionUtils `? I have searched the codebase for usages of the same approach and haven't found anything in common for this. Did you mean some specific examples?

Reply:
I havent looked into the usage too much but it just felt like "we are filtering some methods with annotations on them", seems like that might be handy in future if we "generify" it enough .... up to you.